src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,199 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+                    textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                    partialRange, cancellationToken).ConfigureAwait(false);
+
+                var firstToken = true;
+                var endAbsoluteLine = 0;
+                for (var tokenIndex = 0; tokenIndex < tokens.Data.Length; tokenIndex += 5)
+                {
+                    if (tokenIndex == 0)
+                    {
+                        endAbsoluteLine = tokens.Data[0];
+                    }
+                    else
+                    {
+                        endAbsoluteLine += tokens.Data[tokenIndex];
+                    }
+
+                    // Skip tokens that are out of the range
+                    if (endAbsoluteLine < cachedRange.End.Line || endAbsoluteLine >= range.End.Line)
+                    {
+                        continue;
+                    }
+
+                    if (firstToken)
+                    {
+                        tokenList.Add(endAbsoluteLine - absoluteLine);
+                        firstToken = false;
+                    }
+                    else
+                    {
+                        tokenList.Add(tokens.Data[tokenIndex]);
+                    }
+
+                    tokenList.Add(tokens.Data[tokenIndex + 1]);
+                    tokenList.Add(tokens.Data[tokenIndex + 2]);
+                    tokenList.Add(tokens.Data[tokenIndex + 3]);
+                    tokenList.Add(tokens.Data[tokenIndex + 4]);
+                }
+            }
+
+            return new SemanticTokens { Data = tokenList.ToImmutableArray() };
+
+            async Task<SemanticTokens> ComputeAndCacheTokensAsync(
+                TextDocumentIdentifier textDocumentIdentifier,
+                DocumentSnapshot documentSnapshot,
+                int documentVersion,
+                VersionStamp semanticVersion,
+                Range partialRange,
+                CancellationToken cancellationToken)
+            {
+                var (tokens, isCSharpFinalized) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, partialRange, cancellationToken);
+
+                // Tokens might be null if there are no Razor/C# elements in the range. We don't want this
+                // to prevent the display of the rest of the tokens, so we just assume that range is empty.
+                if (tokens is null)
+                {
+                    tokens = new SemanticTokens { Data = ImmutableArray<int>.Empty };
+                }
+
+                if (isCSharpFinalized)
+                {
+                    _tokensCache.CacheTokens(textDocumentIdentifier.Uri, semanticVersion, partialRange, tokens.Data.ToArray());
+                }
+
+                return tokens;
+            }
         }
 
         private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(DocumentSnapshot documentSnapshot)
```

Comment:
I'm not sure if this is actually the semantic version you want to cache.  Can you verify that if you make a change in a referenced project (that changes semantics in your current document) that it is reflected here?

Reply:
This might be what you want instead
https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/Project.cs,484

Reply:
Yes, we need to cache the Semantic Version of the Project for exactly the reasons @dibarbet implies.

Reply:
I should mention the names of the methods are also slightly confusing - you want [GetDependentVersionAsync](https://sourceroslyn.io/Microsoft.CodeAnalysis.Workspaces/R/051437b1b48636ed.html).

The method [GetDependentSemanticVersionAsync](https://sourceroslyn.io/Microsoft.CodeAnalysis.Workspaces/R/71377eb82e6e8834.html) is a bit different as it only changes when the public semantic version of the project changes.  That means that changes inside a method in your document will not trigger this version to change, so want GetDependentVersionAsync instead.

Reply:
I have confirmed that making a change in a referenced project (Renaming a component) is reflected in coloration changes automatically.

That said I do think you're correct, what we've got will change any time the any part of the document changes. The trick though is that `documentSnapshot.Project` is a `ProjectSnapshot`, not a `Project`. I'm sure there's some way to plumb the correct thing through, but rather than do that in this PR I filed https://github.com/dotnet/razor-tooling/issues/6071.